### PR TITLE
[ENH] Add support for self-shielding in heating/cooling rates calculation for RAMSES

### PR DESCRIPTION
**Pull request by cphyc**
## PR Summary

RAMSES uses some self-shiedling to compute the cooling/heating rates.
This parameter is used in `hydro/cooling_fine.f90`. When self-shielding is activated, a "boost factor" is computed as
boost =exp(-nH/0.01), and the effective density used to look-up the cooling tables is `nH/boost`.

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/frontends/ramses/fields.py
Comment (V-Nathir) [inline yt/frontends/ramses/fields.py:391]:
```suggestion
                # _X * density is not corrected from metallicity. Being _X+_Y+ Z > 1, it has to be normalised. 
                # (1-_Y)*(1-Z) creates a yz parameter that deviates the value apprx >0.2
                Z = data["gas", "metallicity"]
                nH = (1 - _Y - Z)/(_X + _Y + Z) * data["gas", "density"] / mh
```

File: yt/frontends/ramses/data_structures.py
Comment (chrishavlin) [inline yt/frontends/ramses/data_structures.py:1125]:
```suggestion
            nml = self.parameters["namelist"]
```

File: yt/frontends/ramses/data_structures.py
Comment (chrishavlin) [inline yt/frontends/ramses/data_structures.py:802]: Setting this to None will fix the failing `load_archive` test 
```suggestion
    self_shielding: Optional[bool] = None
```

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by cphyc** [comment]
Let's hold this a bit more, we probably need testing before this goes in!

**Comment by cphyc** [comment]
Update: this is almost ready to review. I will add a test once #4817 is merged in.

**Comment by cphyc** [comment]
The error is due to a missing simulation dataset on Jenkins (`ramses_star_formation/output_00013`). @Xarthisius is this something you could do?

**Comment by Xarthisius** [comment]
@yt-fido test this please

File: yt/frontends/ramses/fields.py
Comment (chrishavlin) [inline yt/frontends/ramses/fields.py:424]: `nh` here will have dimensions of of 1/length**3 -- any need to force a particular unit before using the numpy functions below?

File: yt/frontends/ramses/tests/test_outputs.py
Comment (chrishavlin) [inline yt/frontends/ramses/tests/test_outputs.py:692]: seems pretty unlikely it'd happen and cause issues, but this does introduce a race condition -- since this manager actually overwrites the existing namelist file, if tests were run in parallel then other tests using `ramses_new_fromat` could read in the overwritten namelist unintentionally (or catch the file during the active write?). It is a small enough dataset (4.4 MB) that you could instead copy the whole dataset into a temporary directory and modify the namelist files appropriately (which would be cleaner to do with pytest over in `test_outputs_pytest.py`). Would take a bit of extra IO but IMO worth it for the extra future-proofing, but also a small enough thing that I could be convinced that I'm overthinking this :)

**Review by chrishavlin — COMMENTED** [review]
generally LGTM, but left a couple of questions/comments

File: yt/frontends/ramses/data_structures.py
Comment (cphyc) [reply yt/frontends/ramses/data_structures.py:819]: Done

File: yt/frontends/ramses/tests/test_outputs.py
Comment (cphyc) [reply yt/frontends/ramses/tests/test_outputs.py]: This should be good now :)

File: yt/frontends/ramses/tests/test_outputs.py
Comment (chrishavlin) [reply yt/frontends/ramses/tests/test_outputs.py:702]: apologies this turned out to be a little annoying :)

**Comment by cphyc** [comment]
@chrishavlin I am very confused about the nose failure -_-

**Review by chrishavlin — APPROVED** [review]

**Comment by chrishavlin** [comment]
it was actually a trivial merge, so just did it myself :)

**Pull request merged**
